src/main/java/ui/components/pickers/LabelPickerState.java

```diff
@@ -0,0 +1,272 @@
+                    matchedLabels, OptionalInt.of(currentSuggestionIndex.getAsInt() + 1));
+        }
+        return this;
+    }
+
+    public LabelPickerState previousSuggestion() {
+        if (canDecreaseIndex()) {
+            return new LabelPickerState(initialLabels, removedLabels, addedLabels,
+                        matchedLabels, OptionalInt.of(currentSuggestionIndex.getAsInt() - 1));
+        }
+        return this;
+    }
+
+    public LabelPickerState updateMatchedLabels(Set<String> repoLabels, String query) {
+        List<String> newMatchedLabels = convertToList(repoLabels);
+
+        newMatchedLabels = filterByName(newMatchedLabels, getName(query));
+        newMatchedLabels = filterByGroup(newMatchedLabels, getGroup(query));
+
+        OptionalInt newSuggestionIndex;
+        if (query.equals("") || newMatchedLabels.isEmpty()) {
+            newSuggestionIndex = OptionalInt.empty();
+        } else {
+            newSuggestionIndex = OptionalInt.of(0);
+        }
+
+        return new LabelPickerState(initialLabels, addedLabels, removedLabels, newMatchedLabels, newSuggestionIndex);
+    }
+
+    /*
+     * Methods for updating UI
+     */
+
+    public List<String> getAssignedLabels() {
+        List<String> assignedLabels = new ArrayList<>();
+        assignedLabels.addAll(initialLabels);
+        assignedLabels.addAll(addedLabels);
+        assignedLabels.removeAll(removedLabels);
+        return assignedLabels;
+    }
+
+    public List<String> getInitialLabels() {
+        return convertToList(initialLabels);
+    }
+
+    public List<String> getRemovedLabels() {
+        return convertToList(removedLabels);
+    }
+
+    public List<String> getAddedLabels() {
+        return convertToList(addedLabels);
+    }
+
+    public Optional<String> getCurrentSuggestion() {
+        if (currentSuggestionIndex.isPresent() && isValidSuggestionIndex()) {
+            return Optional.of(getSuggestedLabel());
+        }
+        return Optional.empty();
+    }
+
+    public List<String> getMatchedLabels() {
+        return matchedLabels;
+    }
+
+    /*
+     * Helper functions
+     */
+
+    private String getMatchedLabelName(Set<String> repoLabels, String keyword) {
+        List<String> newMatchedLabels = new ArrayList<>();
+        newMatchedLabels.addAll(repoLabels);
+        newMatchedLabels = filterByName(newMatchedLabels, getName(keyword));
+        newMatchedLabels = filterByGroup(newMatchedLabels, getGroup(keyword));
+        return newMatchedLabels.get(0);
+    }
+
+    private boolean hasExactlyOneMatchedLabel(Set<String> repoLabels, String keyword) {
+        List<String> newMatchedLabels = new ArrayList<>();
+        newMatchedLabels.addAll(repoLabels);
+        newMatchedLabels = filterByName(newMatchedLabels, getName(keyword));
+        newMatchedLabels = filterByGroup(newMatchedLabels, getGroup(keyword));
+        return newMatchedLabels.size() == 1;
+    }
+
+    private void removeConflictingLabels(String name) {
```

Comment:
I get what a 'conflicting label' is from reading the method, but you might want to define it in terms of exclusive labels in a comment.